src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs

```diff
@@ -52,67 +76,176 @@ public void AddDependencyLocation(string fullPath)
+        /// </summary>
+        protected string? GetBestPath(AssemblyName assemblyName)
         {
+            if (assemblyName.Name is null)
+            {
+                return null;
+            }
+
+            ImmutableHashSet<string>? paths;
             lock (_guard)
             {
-                _knownAssemblyPathsBySimpleName.TryGetValue(simpleName, out var paths);
-                return paths;
+                if (!_knownAssemblyPathsBySimpleName.TryGetValue(assemblyName.Name, out paths))
+                {
+                    return null;
+                }
             }
+
+            // Sort the candidate paths by ordinal, to ensure determinism with the same inputs if you were to have
+            // multiple assemblies providing the same version.
+            string? bestPath = null;
+            AssemblyName? bestName = null;
+            foreach (var candidateOriginalPath in paths.OrderBy(StringComparer.Ordinal))
+            {
+                (AssemblyName? candidateName, string candidateRealPath) = GetAssemblyInfoForPath(candidateOriginalPath);
+                if (candidateName is null)
+                {
+                    continue;
+                }
+
+                bool isMatch;
+#if NETCOREAPP
+                isMatch = candidateName.Name == assemblyName.Name;
+#else
+                isMatch =
+                    candidateName.Name == assemblyName.Name &&
+                    candidateName.Version >= assemblyName.Version &&
+                    candidateName.GetPublicKeyToken().AsSpan().SequenceEqual(assemblyName.GetPublicKeyToken().AsSpan());
+#endif
+
+                if (isMatch)
+                {
+                    if (candidateName.Version == assemblyName.Version)
+                    {
+                        return candidateRealPath;
+                    }
+
+                    if (bestName is null || candidateName.Version > bestName.Version)
+                    {
+                        bestPath = candidateRealPath;
+                        bestName = candidateName;
+                    }
+                }
+            }
+
+            return bestPath;
         }
 
         /// <summary>
         /// When overridden in a derived class, allows substituting an assembly path after we've
         /// identified the context to load an assembly in, but before the assembly is actually
         /// loaded from disk. This is used to substitute out the original path with the shadow-copied version.
         /// </summary>
-        protected virtual string GetPathToLoad(string fullPath)
+        protected virtual string PreparePathToLoad(string fullPath) => fullPath;
+
+        /// <summary>
+        /// When <see cref="PreparePathToLoad(string)"/> is overriden this returns the most recent
+        /// real path calculated for the <paramref name="originalFullPath"/>
+        /// </summary>
+        internal string GetRealLoadPath(string originalFullPath)
```

Comment:
This method also appears to only be used in tests. Consider adding a comment to that effect.

Reply:
Why though? It's designed to be used in product or test code. I'm not sure the value in commenting it's only used in tests at the moment. It's perfectly valid for product code to depend on in the future.